src/profile-logic/call-node-samples.js

```diff
@@ -0,0 +1,221 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+// @flow
+import type { Thread } from '../types/profile';
+import type { Milliseconds } from '../types/units';
+import type {
+  IndexIntoCallNodeTable,
+  CallNodeTable,
+  CallNodeInfo,
+} from '../types/profile-derived';
+
+import { getCallNodePath } from './profile-data';
+
+export type CallNodeSamplesByDepth = Array<{
+  start: Milliseconds[],
+  end: Milliseconds[],
+  callNode: IndexIntoCallNodeTable[],
+  length: number,
+}>;
+
+type LastSeen = {
+  startTimeByDepth: number[],
+  callNodeIndexByDepth: IndexIntoCallNodeTable[],
+};
+
+type SamplesByFuncPath = {
+  [string]: {
+    samples: number,
+    depth: number,
+    callNodeIndex: IndexIntoCallNodeTable,
+  },
+};
+
+/**
+ * Build a CallNodeSamplesByDepth table from a given thread.
+ *
+ * @param {object} thread - The profile thread.
+ * @param {object} callNodeInfo - from the callNodeInfo selector.
+ * @param {integer} maxDepth - The max depth of the all the stacks.
+ * @return {array} callNodeSamplesByDepth
+ */
+export function getCallNodeSamplesByDepth(
+  thread: Thread,
+  callNodeInfo: CallNodeInfo,
+  maxDepth: number
+): CallNodeSamplesByDepth {
+  const { callNodeTable, stackIndexToCallNodeIndex } = callNodeInfo;
+  const callNodeSamplesByDepth = Array.from({ length: maxDepth + 1 }, () => ({
+    start: [],
+    end: [],
+    callNode: [],
+    length: 0,
+  }));
+
+  const lastSeen: LastSeen = {
+    startTimeByDepth: [],
+    callNodeIndexByDepth: [],
+  };
+
+  const sampleCounts = _collectSamplesPerFuncPath(
+    thread,
+    stackIndexToCallNodeIndex,
+    callNodeTable
+  );
+  const paths = Object.keys(sampleCounts);
+  paths.sort(_compareFuncPath);
+
+  let sampleTime = 0;
+  let previousDepth = -1;
+  for (const path of paths) {
+    const { samples, callNodeIndex, depth } = sampleCounts[path];
+
+    // Find the depth of the nearest shared stack.
+    const depthToPop = _findNearestSharedStackDepth(
+      callNodeTable,
+      callNodeIndex,
+      lastSeen,
+      depth
+    );
+    _popStacks(
+      callNodeSamplesByDepth,
+      lastSeen,
+      depthToPop,
+      previousDepth,
+      sampleTime
+    );
+    _pushStacks(
+      thread,
+      lastSeen,
+      depth,
+      callNodeIndex,
+      callNodeTable,
+      sampleTime
+    );
+    previousDepth = depth;
+    sampleTime += samples;
```

Comment:
We're just counting samples here, so the name `sampleTime` should probably change.